Feat kmp sdk

camka14 opened the pull request:
## What does this PR do?

Adds a KMP SDK

## Test Plan

Run KMPAndroid14Java17Test
Test OAuth2, cookies, and iOS file uploads using custom project in Android Studio

## Related PRs and Issues

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
yes

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added Kotlin Multiplatform (KMP) SDK generation with full project scaffolding (shared, Android, iOS), sample Android app, CI/publish workflows, and packaging templates.
  * Mobile OAuth2 web-auth flows with cookie handling, platform-specific WebAuth integrations, and utilities for URL parsing.
  * Rich client features: chunked file upload with progress, realtime subscriptions, comprehensive JSON/serialization helpers, and platform IO utilities.

* **Tests**
  * Added end-to-end Kotlin tests and PHP test harness for KMP build scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

camka14 committed:
Initial twig translation of KMP SDK

camka14 committed:
Add Test Generation,

camka14 committed:
Fix Test and Release compiling

camka14 committed:
Pass all tests

camka14 committed:
Merge branch 'master' of https://github.com/appwrite/sdk-generator into feat-kmp-sdk

camka14 committed:
Remove Document twig file, remove Document type  form RealtimModels

camka14 committed:
Update x-appwrite-response-format to 1.6.0

elesey committed:
Fix OAuth2 login handling

camka14 committed:
Merge branch 'master' of https://github.com/appwrite/sdk-generator into feat-kmp-sdk

elesey committed:
Switch WebAuthComponent to activity instead of context

elesey committed:
Add OAuth2 handling

camka14 committed:
Merge branch 'master' of https://github.com/appwrite/sdk-generator into feat-kmp-sdk

camka14 committed:
Fix iOS and JVM OAuth2Extensions class names

camka14 committed:
Fix update and create document methods

Certain data classes weren't serializing properly, moved actualSerializer to be above apiParams, added json encoding to data.

camka14 committed:
Fix old cookies being used when domain is preceded by "."

camka14 committed:
Fix socket creation in Realtime.kt

The socket was being created unnecessarily whenever a subscription was removed. This change ensures that the socket is only created when there are active channels.

camka14 committed:
Fix encoding prefs error

abnegate commented on templates/kmp/shared/src/commonMain/kotlin/io/package/extensions/JsonObjectBuilderExtensions.kt.twig:
```diff
@@ -0,0 +1,32 @@
+        is String -> put(key, value)
+        is Number -> put(key, value)
+        is Boolean -> put(key, value)
+        is List<*> -> {
+            val jsonArray = JsonArray(value.map { item ->
+                when (item) {
+                    is String -> JsonPrimitive(item)
+                    is Number -> JsonPrimitive(item)
+                    is Boolean -> JsonPrimitive(item)
+                    is Enum<*> -> JsonPrimitive(item.name)
+                    null -> JsonNull
+                    else -> JsonPrimitive(item.toString())
+                }
+            })
+            put(key, jsonArray)
+        }
+
```
This should recurse, a list could contain a list

abnegate commented on templates/kmp/shared/src/commonMain/kotlin/io/package/services/Realtime.kt.twig:
We have recently added a ping/ping heartbeat for realtime, please check the Android implementation and sync here: https://github.com/appwrite/sdk-for-android/blob/main/library/src/main/java/io/appwrite/services/Realtime.kt

abnegate commented on templates/kmp/shared/src/commonMain/kotlin/io/package/services/Service.kt.twig:
```diff
@@ -0,0 +1,191 @@
+{%~ endif %}
+
+/**
+ * {{ service.description | raw | replace({"\n": "", "\r": ""}) }}
```
We might to replace newlines with spaces instead of empty

abnegate commented on templates/kmp/shared/src/iosMain/kotlin/io/package/Client.ios.kt.twig:
```diff
@@ -0,0 +1,46 @@
+
+    init {
+        httpClient = createHttpClient(selfSigned, iosCookieStorage)
+        headers = mutableMapOf(
+            "content-type" to "application/json",
+            "x-sdk-name" to "{{ sdk.name }}",
+            "x-sdk-platform" to "{{ sdk.platform }}",
+            "x-sdk-language" to "{{ language.name | caseLower }}",
+            "x-sdk-version" to "{{ sdk.version }}"{% if spec.global.defaultHeaders | length > 0 %},{% endif %}
```
Let's add origin and user agent here too

abnegate commented on templates/kmp/shared/src/androidMain/kotlin/io/package/HttpClientConfig.kt.twig:
Let's add the .android in the name here

camka14 replied:
That's normally for expect/actual annotated classes. Since it isn't in the common main I didn't use it. If you still prefer the annotation we can add it, but I don't think that's the standard.

abnegate commented on templates/kmp/shared/src/iosMain/kotlin/io/package/WebAuthComponent.ios.kt.twig:
```diff
@@ -0,0 +1,118 @@
+                return@suspendCancellableCoroutine
+            }
+
+            UIApplication.sharedApplication.openURL(nsUrl, options = emptyMap<Any?, Any>()) { success ->
```
Are we able to use `ASAuthenticationSession` here instead of opening in the browser? You can check the Apple SDK implementation here: https://github.com/appwrite/sdk-for-apple/blob/main/Sources/Appwrite/OAuth/WebAuthComponent.swift

camka14 replied:
Probably, I don't remember why I went with the browser implementation. But this will take me some time.

abnegate commented on tests/KMPAndroid14Java17Test.php:
```diff
@@ -0,0 +1,36 @@
+        'chmod +x tests/sdks/kmp/gradlew',
+    ];
+    protected string $command =
+        'docker run --rm --platform linux/amd64 --network="mockapi" -v $(pwd):/app -w /app/tests/sdks/kmp alvrme/alpine-android:android-34-jdk17 sh -c "./gradlew :shared:testDebugUnitTest --stacktrace -q && cat shared/result.txt"';
```
Do we need to specify platform here? I think this image is multi-arch

abnegate commented on tests/KMPAndroid14Java17Test.php:
Please add tests for iOS and JVM too. We can also add them in .github/workflows/tests.yml so they run for each PR

camka14 commented on src/SDK/Language/Kotlin.php:
```diff
@@ -466,7 +466,7 @@ protected function getReturnType(array $method, array $spec, string $namespace,
         $ret = $this->toPascalCase($method['responseModel']);
 
-        if ($this->hasGenericType($method['responseModel'], $spec)) {
+        if ($this->hasGenericType($method['responseModel'], $spec) && $withGeneric) {
```
It's for situations like this: 
`serializer = io.appwrite.models.Document.serializer(actualSerializer)`

without the skip I would get `serializer = io.appwrite.models.Document<T>.serializer(actualSerializer)`

I just found how it's been done for the same situation:
`{{sdk.namespace | caseDot}}.models.{{ method.responseModel | caseUcfirst }}`

I can change it to match the original method and get rid of this change if you like that better.

abnegate replied:
Yep let's do it so we can handle both generics and non-generics in one shot

camka14 commented on templates/android/library/src/main/java/io/package/WebAuthComponent.kt.twig:
I added the iOS and Android versions of their implementations. At the time I don't think this was working.

camka14 replied:
I just realized this was the Android file, yes you're right. I think I was going between the two files and didn't realize I made changes to it in the first go around.

camka14 commented on templates/kmp/shared/src/androidMain/kotlin/io/package/extensions/OAuth2Extensions.kt.twig:
I can move the file into `models`? But otherwise it needs to be an extension in each of the platforms, since Android requires the `ComponentActivity`.

camka14 commented on src/SDK/Language/KMP.php:
```diff
@@ -0,0 +1,636 @@
+                'destination'   => '.github/workflows/publish.yml',
+                'template'      => '/kmp/.github/workflows/publish.yml',
+            ],
+//            [
+//                'scope'         => 'method',
+//                'destination'   => 'docs/examples/kotlin/{{service.name | caseLower}}/{{method.name | caseDash}}.md',
+//                'template'      => '/kmp/docs/kotlin/example.md.twig',
+//            ],
+//            [
+//                'scope'         => 'method',
+//                'destination'   => 'docs/examples/java/{{service.name | caseLower}}/{{method.name | caseDash}}.md',
+//                'template'      => '/kmp/docs/java/example.md.twig',
+//            ],
```
Probably so I don't forget that they don't exist yet.

abnegate replied:
Let's uncomment and check out if the docs are generating nicely

camka14 commented on templates/kmp/shared/src/androidMain/kotlin/io/package/extensions/OAuth2Extensions.kt.twig:
```diff
@@ -0,0 +1,79 @@
+import {{ sdk.namespace | caseDot }}.exceptions.AppwriteException
+import {{ sdk.namespace | caseDot }}.webInterface.UrlParser
+import io.ktor.http.Url
+import kotlinx.coroutines.runBlocking
+
+{% for service in spec | webAuthServices %}
+{% for method in service.methods %}
+suspend fun {{ sdk.namespace | caseDot }}.services.{{ service.className | caseUcfirst }}.{{ method.methodName }}(
```
I remember trying to find a way to do that, but since Android requires the `ComponentActivity` it pretty much requires that I make an extension function in each of the platforms. Otherwise I would make it an expect/actual, but functions can't have different parameter inputs unlike classes.

camka14 commented on templates/kmp/shared/src/androidMain/kotlin/io/package/models/InputFile.android.kt.twig:
In the commonMain. The mimeType just needs a platform specific solution.

camka14 commented on templates/kmp/shared/src/commonMain/kotlin/io/package/serializers/StringCollectionSeriailizer.kt.twig:
It's been a hot minute since I looked at this so I may be wrong. But, I believe it's because of this in the `DynamicLookupSerializer`:
`ContextualSerializer(Any::class, null, emptyArray()).descriptor`

It specifies the `Any` class, so it won't work on a class that is specified as Collection<String>. There might be a way to make the dynamic serializer work more generically, but I think it's more effort than it's worth.

camka14 committed:
Merge branch 'master' of https://github.com/appwrite/sdk-generator into feat-kmp-sdk

camka14 commented on templates/php/tests/IDTest.php.twig:
Nothing, I think they got caught in me updating the line endings. I'll revert them

camka14 commented on templates/kmp/shared/src/androidMain/kotlin/io/package/HttpClientConfig.kt.twig:
```diff
@@ -0,0 +1,63 @@
+                hostnameVerifier { _, _ -> true }
+            }
+        }
+    } else {
+        engine {
+            config {
+                val trustManagerFactory =
+                    TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
+                trustManagerFactory.init(null as KeyStore?)
+                val sslContext = SSLContext.getInstance("TLS").apply {
+                    init(null, trustManagerFactory.trustManagers, SecureRandom())
+                }
+                sslSocketFactory(
+                    sslContext.socketFactory,
+                    trustManagerFactory.trustManagers[0] as X509TrustManager
+                )
+            }
+        }
```
I'm not super familiar with certificates, but from my understanding this uses a PKIX algorithm instead of Trust All. I'm not sure if that's the correct approach, but I figured if there was an option to not run `setSelfSigned` then it should use a standard algorithm?

abnegate replied:
We can remove the else block, since the behavior when not setting self-signed will be the same with or without it 👌

camka14 committed:
PR Changes per comments

camka14 committed:
Revert uninteded changes

camka14 committed:
Revert Uninteded File Change

camka14 committed:
Replace io.appwrite with proper value

camka14 committed:
Update KMP tests, replace spec.namespace with sdk.namespace where needed, Fix http/https check in Client

elesey committed:
Refactor iOS OAuth2 authentication

This commit refactors the OAuth2 authentication flow on iOS to use `ASWebAuthenticationSession` instead of `UIApplication.openURL`. This provides a more secure and user-friendly authentication experience.

Key changes:
- `WebAuthComponent.ios.kt`:
    - Replaced `UIApplication.openURL` with `ASWebAuthenticationSession`.
    - Added `PresentationContextProviderImpl` to provide a presentation anchor for the authentication session.
    - Updated `PendingAuth` to store the `ASWebAuthenticationSession`.
    - Improved error handling and cancellation.
    - Added `setCookieStorage` to allow passing the cookie storage instance.
    - Modified `handleIncomingCookie` to parse cookie details from the callback URL and store them using the provided `IosCookieStorage` and `NSUserDefaults`.
    - Added a `cleanup` method to cancel pending authentication sessions.
- `OAuth2Extensions.kt`:
    - Updated `createOAuth2Session` and `createOAuth2Token` to use the new `WebAuthComponent` and pass the `iosCookieStorage`.
    - Ensured cookies are correctly constructed and stored.
    - Improved error handling for missing authentication cookies.

elesey committed:
Remove println in iosCookieStorage

camka14 committed:
Incorrect new line

camka14 committed:
Added variables for ios client headers

camka14 committed:
AppwriteException incorrect Case

camka14 committed:
Replace value with variable

camka14 committed:
Fix error thrown outside of scope in iOS OAuth Cancellation

camka14 committed:
Fix iOS cookie storage not recreating cookies properly

abnegate commented on templates/kmp/README.md.twig:
```diff
@@ -0,0 +1,79 @@
+   sourceSets {
+      commonMain {
+         dependencies {
+            implementation("io.github.camka14.appwrite:sdk-for-kmp:0.1.0")
```
Let's template the package name here

camka14 committed:
Merge branch 'feat-kmp-sdk' of https://github.com/camka14/sdk-generator into feat-kmp-sdk

camka14 committed:
Merge branch 'master' of https://github.com/appwrite/sdk-generator into feat-kmp-sdk

camka14 committed:
Update dependencies and improve ID generation logic

camka14 committed:
Enhance Realtime service with heartbeat mechanism and nullable data handling

camka14 committed:
Merge commit '1a7a3b89147aa8c1bde5247f8eeb7e4832c6016d'

camka14 commented on templates/kmp/shared/src/iosMain/kotlin/io/package/cookies/IosCookieStorage.kt.twig:
```diff
@@ -0,0 +1,243 @@
+            NSHTTPCookieValue to storedCookie.value,
+            NSHTTPCookiePath to storedCookie.path,
+            NSHTTPCookieDomain to storedCookie.domain
+        )
```
I'm not able to find an option for samesite in both the iOS property names and ktor cookie class, I also don't see it being handled in the swift library. Is samesite generally handled in some other way?

camka14 replied:
I've updated to handle the other 4 properties

camka14 committed:
Refactor OAuth2 authentication flow and remove unused JVM components

- Updated OAuth2Extensions.kt.twig to use CoroutineScope for asynchronous operations during OAuth authentication.
- Removed AllCertsTrustManager.kt.twig, Client.jvm.kt.twig, HttpClient.kt.twig, WebAuthComponent.jvm.kt.twig, and related files as they are no longer needed.
- Cleaned up OAuth2Extensions.kt.twig by removing redundant code and improving error handling.
- Removed InputFile.jvm.kt.twig, JvmParsedUrl.kt.twig, and UrlParser.jvm.kt.twig as part of the cleanup.
- Updated test files for Android and KMP to reflect changes in the SDK structure and ensure compatibility with the latest changes.

camka14 commented:
> Any news when the sdk will be available?

In the mean time you can use my sdk here: https://github.com/camka14/sdk-for-kmp

camka14 committed:
Refactor KMP SDK templates and improve AndroidManifest.xml configuration

camka14 committed:
Merge branch 'master' into feat-kmp-sdk

camka14 commented on templates/kmp/shared/src/iosMain/kotlin/io/package/HttpClientConfig.ios.kt.twig:
```diff
@@ -0,0 +1,58 @@
+                completionHandler(disposition.convert(), null)
+            }
+        }
+    }
```
Removed

camka14 commented on templates/kmp/shared/src/androidMain/kotlin/io/package/HttpClientConfig.kt.twig:
```diff
@@ -0,0 +1,80 @@
+        storage = dataStoreCookieStorage
+    }
+    install(WebSockets) {
+        pingInterval = 30.seconds
```
It does not, I went ahead and removed it to match the android behavior.

camka14 commented on templates/kmp/androidApp/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,3 @@
+<resources>
+    <string name="app_name">Appwrite KMP SDK</string>
```
Made the file into a template and templated, although I noticed these files aren't templated in the Android sdk?

camka14 committed:
Refactor Client and WebAuthComponent to improve context handling and cookie storage assignment

camka14 committed:
Refactor KMP SDK templates and improve cookie handling in WebAuthComponent

camka14 committed:
Enhance Query class with additional filtering methods and streamline existing function implementations

camka14 committed:
Add default CoroutineScope creation in WebAuthComponent for improved cookie handling
